### PR TITLE
Add deployment settings tests, cleanup and better handling of non interactive sessions

### PR DESCRIPTION
**Pull request by glena**
Add deployment settings tests, cleanup and better handling of non interactive executions

Fix https://github.com/pulumi/pulumi-service/issues/20781

### CODE REVIEW

File: pkg/cmd/pulumi/deployment_settings_config.go
Comment (tgummerer) [inline pkg/cmd/pulumi/deployment_settings_config.go:526]: Don't we have a function that's not user skippable?  Since this no longer seems to be skippable by the user, so using this function is kinda confusing.  Or should we pass `!d.Interactive` below instead of `false`?

**Review by tgummerer — COMMENTED** [review]
https://github.com/pulumi/pulumi/pull/16677/files#diff-3349ff3e731016b1b6ba4aeb9f3d054767a3e0f96c65608165da2c11167eb56bR134 this condition should be removed now I think?   Since we no longer rely on `--yes` making the command not prompt for things?

File: pkg/cmd/pulumi/deployment_settings_config.go
Comment (glena) [reply pkg/cmd/pulumi/deployment_settings_config.go:526]: They are still skippable, just not managed by the user. In this case there is no way for the user to reach out this point in a non interactive session, that is why I hardcoded it to false. Now that I am writing this, it should not use the skippable prompt

Comment (glena) [reply pkg/cmd/pulumi/deployment_settings_config.go]: Done both comments

**Review by tgummerer — APPROVED** [review]
This PR was quite a lot to get through.  For another time it would be nice to have it split up into at least 3 parts:
- adding the tests
- cleanups
- better handling of non-interactive sessions

**Review by mikhailshilkov — CHANGES_REQUESTED** [review]
Please don't merge it yet, we want to give this another review, thank you!

**Comment by mikhailshilkov** [comment]
I dismissed my review so that the PR is not blocked on me personally, but please get an approve from @pulumi/platform

**Comment by glena** [comment]
@mikhailshilkov it is already approved by @tgummerer, is it ok or should I ask for another round?

**Review by lunaris — APPROVED** [review]

**Comment by tgummerer** [comment]
@glena I removed this PR from the queue because some tests are failing on windows:

```
=== FAIL: cmd/pulumi TestValidateRelativeDirectory (0.52s)
    deployment_settings_utils_test.go:100: 
        	Error Trace:	D:/a/pulumi/pulumi/pkg/cmd/pulumi/deployment_settings_utils_test.go:100
        	Error:      	Received unexpected error:
        	            	invalid relative path ./goproj
        	Test:       	TestValidateRelativeDirectory

=== FAIL: cmd/pulumi TestRepoLookup/should_handle_directories_that_are_a_git_repo (0.47s)
    deployment_settings_utils_test.go:72: 
        	Error Trace:	D:/a/pulumi/pulumi/pkg/cmd/pulumi/deployment_settings_utils_test.go:72
        	Error:      	Not equal: 
        	            	expected: "goproj/something"
        	            	actual  : "goproj\\something"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-goproj/something
        	            	+goproj\something
        	Test:       	TestRepoLookup/should_handle_directories_that_are_a_git_repo
    deployment_settings_utils_test.go:82: 
        	Error Trace:	D:/a/pulumi/pulumi/pkg/cmd/pulumi/deployment_settings_utils_test.go:82
        	Error:      	Not equal: 
        	            	expected: "."
        	            	actual  : "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\pulumi_auto4118040056"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-.
        	            	+C:\Users\RUNNER~1\AppData\Local\Temp\pulumi_auto4118040056
        	Test:       	TestRepoLookup/should_handle_directories_that_are_a_git_repo

=== FAIL: cmd/pulumi TestRepoLookup (0.00s)
```

You can have CI run those tests on the PR as well by applying the `ci/test` label.

**Comment by glena** [comment]
@tgummerer oh are those not run automatically on every build? thanks for the heads up

**Comment by tgummerer** [comment]
> @tgummerer oh are those not run automatically on every build? thanks for the heads up

No, Windows and MacOS tests are currently only run in the merge queue by default.  They are run in the PR if the `ci/test` label is applied before the tests are being run.

**Comment by glena** [comment]
@tgummerer I couldnt look into this earlier beause I was on oncall duties. The PR should be ready now, I fixed the path handling (I realize I was using the path module instead of filepath, plus I stopped using `..` to get to the parent directory) in the tests and this helper me realise the repo dir should be converted always to use the linux path separator (as deployments always run in a linux container). Could you take a final look to the changes?

File: pkg/cmd/pulumi/deployment_settings_utils.go
Comment (tgummerer) [inline pkg/cmd/pulumi/deployment_settings_utils.go:241]: This is a bit pedantic, but I think this should be called `convertToUnixPath`.

File: pkg/cmd/pulumi/deployment_settings_utils.go
Comment (tgummerer) [inline pkg/cmd/pulumi/deployment_settings_utils.go:143]: I'm not sure I quite understand why this should be converted to a Unix path.  I understand deployments is always using Linux containers, but if I'm reading things correctly, it always points to a *local* directory anyway, not the remote one?  Also the user is prompted for a value for the repository root, which they may or may not enter in the native or Unix name, so converting to the Unix path should be happening *after* that in any case.  Am I missing something here?  

As an aside, all the deployment settings struct could use some more documentation.

File: pkg/cmd/pulumi/deployment_settings_utils.go
Comment (glena) [reply pkg/cmd/pulumi/deployment_settings_utils.go:241]: sure

File: pkg/cmd/pulumi/deployment_settings_utils.go
Comment (glena) [reply pkg/cmd/pulumi/deployment_settings_utils.go:143]: > it always points to a local directory anyway
yes and that configuration applies when running the deployment, which runs in a unix environment, hence the need to use the appropriate path separator

> so converting to the Unix path should be happening after that in any case
makes sense, my intent was to keep this input consistent across the different interfaces where it is expected to be "unix format". But makes sense to keep it consistent with the environment it is running too, I will change

Comment (glena) [reply pkg/cmd/pulumi/deployment_settings_utils.go]: done, I also remove the two "convert path" functions and used filepath's ToSlash and FromSlash implementation

**Pull request merged**
